File: android/src/main/java/dev/harding/plugins/launchnative/LaunchNativePlugin.java
package dev.harding.plugins.launchnative;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "LaunchNative")
public class LaunchNativePlugin extends Plugin {

    private LaunchNative implementation = new LaunchNative();

    @PluginMethod
    public void attempt(PluginCall call) {
        String url = call.getString("url");

        JSObject ret = new JSObject();
        ret.put("completed", implementation.attempt(getContext(), url));
        call.resolve(ret);
    }

}

File: ios/Sources/LaunchNativePlugin/LaunchNativePlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(LaunchNativePlugin)
public class LaunchNativePlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "LaunchNativePlugin"
    public let jsName = "LaunchNative"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "attempt", returnType: CAPPluginReturnPromise),
    ]
    @objc func attempt(_ call: CAPPluginCall) {
        let url = call.getString("url") ?? ""

        guard let parsedUrl = URL(string: url) else {
            call.resolve([
                "completed": false
            ])
            return
        }

        DispatchQueue.main.async {
            UIApplication.shared.open(parsedUrl, options: [.universalLinksOnly: true]) { success in
                if success {
                    call.resolve([
                        "completed": true
                    ])
                } else {
                    call.resolve([
                        "completed": false
                    ])
                }
            }
        }
    }
}
